File: cpp-07/ex0/CreatureCard.cpp
//
// Created by robot on 3/9/26.
//

#include "CreatureCard.hpp"

std::map<std::string, std::variant<std::string, int>>
CreatureCard::play(std::map<std::string, std::variant<std::string, int>> game_state)
{
    std::map<std::string, std::variant<std::string, int>> dict = {
            {"card_played", name},
            {"mana_used", cost},
            {"effect", std::string("Creature summoned to battlefield")}
    };
    return dict;
}

std::map<std::string, std::variant<std::string, int>> CreatureCard::get_card_info() const
{
    std::map<std::string, std::variant<std::string, int>> info = Card::get_card_info();
    info["type"] = std::string("Creature");
    info["attack"] = attack;
    info["health"] = health;
    return info;
}

std::map<std::string, std::variant<std::string, int>> CreatureCard::attack_target(const std::string &target_name)
{
    std::map<std::string, std::variant<std::string, int>> dict = {
            {"attacker", name},
            {"target", target_name},
            {"damage_dealt", attack},
            {"combat_resolved", 1} // Using 1 for True as bool isn't in variant here
    };
    return dict;
}

File: python-07/ex0/CreatureCard.py
# flake8: noqa
from ex0.Card import Card


class CreatureCard(Card):
    def __init__(self, name: str, cost: int, rarity: str, attack: int, health: int):
        super().__init__(name, cost, rarity)
        self.attack = attack
        self.health = health

    def play(self, game_state: dict) -> dict:
        self.game_state = game_state
        game_state = {
                      "card_played": self.name,
                      "mana_used": self.cost,
                      "effect": "Creature summoned to battlefield"
                      }
        return game_state

    def get_card_info(self) -> dict:

        return {
                    "name": self.name,
                    "cost": self.cost,
                    "rarity": self.rarity,
                    "type": "Creature",
                    "attack": self.attack,
                    "health": self.health
                }

    def is_playable(self, available_mana: int) -> bool:
        self.available_mana = available_mana
        if available_mana >= self.cost:
            return True
        return False

    def attack_target(self, target) -> dict:
        self.target = target
        return {"attacker": self.name,
                "target": target,
                "combat_resolved": True
                }

    def is_valid(self) -> bool:
        self.available_mana -= self.cost
        if self.available_mana < self.cost:
            return self.is_playable(self.available_mana)
        if self.attack and self.health < 0:
            return False
